core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java

```diff
@@ -120,24 +121,23 @@ public static EnumerableHashJoin create(
       }
     }
 
+    RelOptCostFactory costFactory = planner.getCostFactory();
     // Cheaper if the smaller number of rows is coming from the LHS.
     // Model this by adding L log L to the cost.
     final double rightRowCount = right.estimateRowCount(mq);
     final double leftRowCount = left.estimateRowCount(mq);
-    if (Double.isInfinite(leftRowCount)) {
-      rowCount = leftRowCount;
-    } else {
-      rowCount += Util.nLogN(leftRowCount);
-    }
-    if (Double.isInfinite(rightRowCount)) {
-      rowCount = rightRowCount;
-    } else {
-      rowCount += rightRowCount;
+    if (Double.isInfinite(leftRowCount) || Double.isInfinite(rightRowCount)) {
+      return costFactory.makeInfiniteCost();
     }
     if (isSemiJoin()) {
-      return planner.getCostFactory().makeCost(rowCount, 0, 0).multiplyBy(.01d);
+      // For semi- and anti- the right input is hashed
+      rowCount += rightRowCount * 1.1d;
+      return costFactory.makeCost(rowCount, 0, 0).multiplyBy(.01d);
     } else {
-      return planner.getCostFactory().makeCost(rowCount, 0, 0);
+      // Lots of plans assumed the left join was hashed, so
+      // we adjust a cost accordingly
+      rowCount += leftRowCount * 1.1d;
+      return costFactory.makeCost(rowCount, 0, 0);
```

Comment:
Why a hash cost is `row_cnt * 1.1` ?

Reply:
The idea was hashtable creation has a linear cost, but it is a bit more than just processing the entries.

In practice, this needs adjustments, as org.apache.calcite.test.JdbcTest#testAlmostBushy plan must be flipped, but currently, it does not flip, so hash join should be adjusted somehow.

Probably I'm missing hash lookup cost.

Reply:
The problem here is not 1.1, but the problem is join cardinality estimation.
Calcite tends to over-estimate the join cardinality, thus leftRowCount and rightRowCount tend to be noise when compared with the estimated join cardinality, thus all those `+= leftRowCount * 1.1d` change nothing.

So a series of patches (https://issues.apache.org/jira/browse/CALCITE-3666, https://issues.apache.org/jira/browse/CALCITE-3661) is required before this costing could provide any meaningful results :(